Correct compilation errors reported by GCC 10/11

szubersk opened the pull request:
### Motivation and Context
#12167

### Description
New `zfs_type_t` value `ZFS_TYPE_INVALID` is introduced.
Variable initialization is now possible to make GCC happy.

Close #12167.

### How Has This Been Tested?
`$ make`

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

behlendorf commented:
This breaks the build on FreeBSD, 

```sh
  CC       os/freebsd/libzfs_zmount.lo
os/freebsd/libzfs_compat.c:295:10: error: enumeration value 'ZFS_TYPE_INVALID' not handled in switch [-Werror,-Wswitch]
        switch (zhp->zfs_type) {
```

szubersk commented:
> This breaks the build on FreeBSD,


Thanks for spotting this! Corrected and rebased. As a bonus, I got rid of shadowed `errbuf` stack variable in `lib/libzfs/os/freebsd/libzfs_compat.c` (cppcheck complained).

behlendorf commented on module/zfs/zcp_get.c:
```diff
@@ -497,8 +496,8 @@ get_zap_prop(lua_State *state, dsl_dataset_t *ds, zfs_prop_t zfs_prop)
 {
-	int error;
-	zfs_type_t zfs_type;
+	int error = 0;
```
nit: It looks like you could declare `error` below in the one place it's used.

```c
	int error = get_objset_type(ds, &zfs_type);
```

szubersk replied:
Good point!

behlendorf reviewed (approved):
Looks good, thanks for sorting this this.

szubersk commented:
@behlendorf, rebased.

szubersk commented:
@behlendorf I found 2 more issues: one from Debian bullseye ASan+UBSan, the other one from Debian Bookworm. Most likely other would hit them soon.
